tools/test_xenial_upgrade.sh

```diff
@@ -0,0 +1,219 @@
+    $tool exec $name -- sudo apt install ubuntu-advantage-tools -y
+}
+
+function attach_ua() {
+    tool=$1
+    set +x
+    echo "+ $tool exec $name -- sudo ua attach \$UACLIENT_BEHAVE_CONTRACT_TOKEN"
+    $tool exec $name -- sudo ua attach $UACLIENT_BEHAVE_CONTRACT_TOKEN
+    set -x
+}
+
+
+
+
+function test_upgrade_in_container() {
+    setup_container
+
+    h1 "Upgrade to UA 27 while unattached"
+
+    install_new_ua lxc
+
+    h2 "Check for leftover files from old version - verify nothing unexpected is left behind"
+    set +e
+    lxc exec $name -- ls -l /etc/update-motd.d/99-esm /usr/share/keyrings/ubuntu-esm-keyring.gpg /usr/share/keyrings/ubuntu-fips-keyring.gpg /usr/share/man/man1/ubuntu-advantage.1.gz /usr/share/doc/ubuntu-advantage-tools/copyright /usr/share/doc/ubuntu-advantage-tools/changelog.gz /usr/bin/ubuntu-advantage
+    set -e
+
+    h2 "New Status - verify esm-infra available but not enabled; esm-apps not visible"
+    lxc exec $name -- ua status
+
+    h2 "Attach - verify esm-infra automatically enabled; esm-apps not visible"
+    attach_ua lxc
+
+    h2 "Detaching before destruction"
+    lxc exec $name -- ua detach --assume-yes
+
+    teardown_container
+}
+
+
+function test_upgrade_with_livepatch_in_vm() {
+
+    setup_vm
+
+    h1 "Upgrade to UA 27 while old version has livepatch enabled"
+
+    h2 "Enable livepatch on old version"
+    set +x
+    echo "+ multipass exec $name -- ubuntu-advantage enable-livepatch \$LIVEPATCH_TOKEN"
+    multipass exec $name -- sudo ubuntu-advantage enable-livepatch $LIVEPATCH_TOKEN
+    set -x
+
+    h2 "Status before - old UA and livepatch say enabled"
+    multipass exec $name -- sudo canonical-livepatch status
+    multipass exec $name -- sudo ubuntu-advantage status
+
+    install_new_ua multipass
+
+    h2 "Status after upgrade - livepatch still enabled but new UA doesn't report it"
+    multipass exec $name -- sudo canonical-livepatch status
+    multipass exec $name -- sudo ua status
+
+    h2 "Attach - verify that livepatch is disabled and re-enabled"
+    attach_ua multipass
+
+    h2 "Status after attach - both livepatch and UA should say enabled"
+    multipass exec $name -- sudo canonical-livepatch status
+    multipass exec $name -- sudo ua status
+
+    h2 "Detaching before destruction"
+    multipass exec $name -- sudo ua detach --assume-yes
+
+    teardown_vm
+}
+
+function test_upgrade_with_fips_in_vm() {
+
+    setup_vm
+
+    h1 "Upgrade to UA 27 while old version has fips enabled"
+
+    h2 "Manual fips check says disabled (file doesn't exist)"
+    set +e
+    multipass exec $name -- sudo cat /proc/sys/crypto/fips_enabled
+    set -e
+
+    h2 "Enable fips on old version"
+    set +x
+    echo "+ multipass exec $name -- ubuntu-advantage enable-fips \$FIPS_CREDS"
+    multipass exec $name -- sudo ubuntu-advantage enable-fips $FIPS_CREDS
+    set -x
+
+    h2 "Reboot to finish fips activation"
+    set +e
+    multipass exec $name -- sudo reboot
+    set -e
```

Comment:
You don't need the set +e -e bookends if you do this:
```suggestion
    multipass exec $name -- sudo reboot || true
```

Reply:
Right! Good tip